Update to GKL 0.4.1, HTSJDK 2.9.1, and enable IntelInflater

gspowley opened the pull request:
Update GKL to 0.4.1 and HTSJDK to 2.9.1.

GKL 0.4.1 provides the following improvements over GKL 0.1.2 (currently used by GATK4):
- Improved performance and compression ratio for level-1 compression
- Improved performance for compression levels > 1
- Decompression acceleration with `IntelInflater`

The `IntelInflater` is enabled by default and can be disabled with the option `--use_jdk_inflater`.

Resolves #2421, #2315, #2302

droazen commented:
@gspowley Looks like integration tests are failing?

gspowley commented:
Yes, debugging that. I'll post my findings.

lbergelson commented:
I can't tell what's wrong, but it's failing because something caught a SIGABRT

gspowley commented:
We found an issue with the way GKL was freeing an internal data structure. The issue was exposed by the `IntelInflaterIntegrationTest`, which uses the Inflater API in a different way than HTSJDK and GATK.

We've fixed that issue and the GATK integration tests pass. We're releasing GKL 0.4.1 now and will update this PR when 0.4.1 is available in Maven Central.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -161,6 +169,8 @@ public Object instanceMainPostParseArgs() {
                 final boolean usingIntelDeflater = (BlockCompressedOutputStream.getDefaultDeflaterFactory() instanceof IntelDeflaterFactory && ((IntelDeflaterFactory)BlockCompressedOutputStream.getDefaultDeflaterFactory()).usingIntelDeflater());
                 logger.info("Deflater " + (usingIntelDeflater ? "IntelDeflater": "JdkDeflater"));
+                final boolean usingIntelInflater = (BlockGunzipper.getDefaultInflaterFactory() instanceof IntelInflaterFactory && ((IntelInflaterFactory)BlockGunzipper.getDefaultInflaterFactory()).usingIntelInflater());
+                logger.info("Inflater " + (usingIntelInflater ? "IntelInflater": "JdkInflater"));
```
What happens if the user requests the intel inflater and it fails to load? Is there a user-friendly log message in that case?

gspowley replied:
Yes, `IntelInflaterFactory` generates a warning:

```java
        logger.warn("IntelInflater is not supported, using Java.util.zip.Inflater");
```

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelDeflaterIntegrationTest.java:
```diff
@@ -94,64 +94,4 @@ public void deflateInflateWithIntelDeflater() {
     }
-
-    @Test(enabled = false)
-    public void bamWriteReadWithIntelDeflater() throws IOException {
```
Can you explain why this test was deleted, and why it was marked as `enabled = false`?

gspowley replied:
This test was deleted here because we already run this test as an integration test in gkl. I believe it was disabled here because of issues running on architectures that do not support `IntelInflater`. We can add a bam based test to gatk.

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelDeflaterIntegrationTest.java:
```diff
@@ -42,7 +42,7 @@ public void testIntelDeflaterIsAvailable(){
             throw new SkipException("IntelDeflater not available for this architecture");
```
The `INPUT_FILE` declared above does not exist (`src/test/resources/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam`), though there is a file with that name in `src/test/resources/large`

gspowley replied:
Oh, the `INPUT_FILE` variable is not used. I'll fix that if we decide to use that bam for the end to end test.

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+public class IntelInflaterIntegrationTest extends BaseTest {
+
+    private final static Logger log = LogManager.getLogger(IntelInflaterIntegrationTest.class);
+    private static final String INPUT_FILE = publicTestDir + "CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam";
```
Same comment here -- `INPUT_FILE` as declared above does not exist (`src/test/resources/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam`), though there is a file with that name in `src/test/resources/large`

gspowley replied:
Same reply :)

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Test
+    public void inflateInflateWithIntelInflater() {
```
`inflateInflate`?

gspowley replied:
Looks like an artifact of search and replace, I'll fix that.

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+            int compressedBytes = 0;
+            while (!deflater.finished()) {
+                compressedBytes = deflater.deflate(compressed, 0, compressed.length);
+            }
```
Will this loop ever iterate more than once? It seems like each pass through the loop would overwrite the previous pass's results.

gspowley replied:
Right, it should not iterate more than once, the test would fail if it did. The test could be made more robust to take care of that condition, or we can move to an end to end bam test.

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+            deflater.end();
+
+            // log results
+            log.info("%d bytes compressed to %d bytes : %2.2f%% compression\n",
```
This should be debug level output

gspowley replied:
Sure, we'll change to debug level or remove this output. This message was useful in GKL testing, but may not be interesting here.

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    @Test
+    public void inflateInflateWithIntelInflater() {
+        if (!isIntelInflaterSupported()) {
```
This test seems fine, but can you also add a test case that does a roundtrip read/write of a bam using the intel inflater?

gspowley replied:
Yes, I'll add a read/write bam test and update the PR when it's ready.

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+                inflater.inflate(result, 0, result.length);
+                inflater.end();
+            } catch (java.util.zip.DataFormatException e) {
+                e.printStackTrace();
```
Should we be trapping this exception like this? Don't we want to let the exception propagate so that the test will faill in this case?

gspowley replied:
Right, we should not catch this exception. That was a debug hook from our original test in GKL. I'll remove the try/catch block.

droazen commented:
First-pass review complete, back to @gspowley to address comments.

gspowley commented:
Added a `PrintReads` based test (`IntelInflaterIntegrationTest.testIntelInflaterDeflaterWithPrintReads`) to test integration of `IntelInflater` and `IntelDeflater`. Removed the previous `IntelInflater` and `IntelDeflater` integration tests, which were basically copies of GKL unit tests. 

**Note:** 
The `PrintReads` test is using `CEUTrio.HiSeq.WGS.b37.NA12878.20.21.tiny.md.bam` for input.

When using `CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam` for input, we see the exception below when comparing `PrintReads` input and output. **This is true when using `IntelInflater`/`IntelDeflater`, as well as JDK `Inflater`/`Deflater`.** Is this a problem?

```
htsjdk.samtools.SAMFormatException: SAM validation error: ERROR: Record 9762, Read name 20GAVAAXX100126:7:2:8126:115177, bin field of BAM record does not equal value computed based on alignment start and end, and length of sequence to which read is aligned
```

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelDeflaterIntegrationTest.java:
```diff
@@ -42,116 +27,7 @@ public void testIntelDeflaterIsAvailable(){
     }
-
-    @Test
-    public void deflateInflateWithIntelDeflater() {
```
This test still seems valuable to me to have in the GATK (`deflateInflateWithIntelDeflater`), even if it is similar to a test in the GKL. Can you restore it?

gspowley replied:
yes, i restored it and addressed the previous comments about this test

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    public void testIntelInflaterDeflaterWithPrintReads() throws Exception {
```
Can you add explicit args for `--use_jdk_inflater` and `--use_jdk_deflater`  here to guarantee that the intel inflater is on for this test, even if the defaults ever were to change?

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Test
+    public void testIntelInflaterDeflaterWithPrintReads() throws Exception {
+        if (!isIntelInflaterSupported()) {
```
Also, can you add two additional test cases: JDK inflater + Intel Deflater, and Intel inflater + JDK deflater, using the same bam? Probably makes sense to add a `DataProvider` that passes in booleans for `--use_jdk_inflater` and `--use_jdk_deflater`.

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Test that it's possible to load libIntelInflater
+ */
+public class IntelInflaterIntegrationTest extends CommandLineProgramTest {
```
Since many of the tests test both the inflater and deflater at once, recommend combining the two test classes into a single `IntelInflaterDeflaterIntegrationTest` class.

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelDeflaterIntegrationTest.java:
```diff
@@ -42,116 +27,7 @@ public void testIntelDeflaterIsAvailable(){
-                     LEN, compressedBytes, 100.0 - 100.0 * compressedBytes / LEN);
-
-            // decompress and check output == input
-            Inflater inflater = new Inflater(true);
```
When you restore this test, recommend having it use an intel inflater as well.

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class IntelInflaterIntegrationTest extends CommandLineProgramTest {
+
+    private static final String INPUT_DIR = publicTestDir + "/large";
```
This is inherited from `BaseTest` as `largeFileTestDir`, so you should use that.

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterIntegrationTest.java:
```diff
@@ -0,0 +1,59 @@
+        args.add("--output"); args.add(outFile.getAbsolutePath());
+
+        // run PrintReads, which uses IntelInflater and IntelDeflater by default
+        new Main().instanceMain(makeCommandLineArgs(args, "PrintReads"));
```
Just call the inherited method `runCommandLine(args)` instead

gspowley replied:
Ah, that's the trick I was looking for.

droazen commented:
Can you run the GATK tool `ValidateSamFile` on `CEUTrio.HiSeq.WGS.b37.NA12878.20.21.tiny.md.bam`, and see if you get the same validation error as you do after roundtripping through the inflater/deflater, @gspowley?

droazen commented:
Second-pass review complete, back to @gspowley

gspowley commented:
Yes, I see the same errors in `CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam` from `ValidateSamFile`.

Here's some of the output:

```
[March 9, 2017 7:03:42 PM EST] org.broadinstitute.hellbender.tools.picard.sam.ValidateSamFile  --input src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam --use_jdk_deflater true --use_jdk_inflater true  --MODE VERBOSE --MAX_OUTPUT 100 --IGNORE_WARNINGS false --VALIDATE_INDEX true --IS_BISULFITE_SEQUENCED false --MAX_OPEN_TEMP_FILES 8000 --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 1 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --help false --version false --verbosity INFO --QUIET false
[March 9, 2017 7:03:42 PM EST] Executing as gspowley@dna on Linux 3.10.0-514.10.2.el7.x86_64 amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_111-b14; Version: Version:4.alpha.2-170-g8d06823-SNAPSHOT
19:03:42.998 INFO  ValidateSamFile - Defaults.BUFFER_SIZE : 131072
19:03:42.999 INFO  ValidateSamFile - Defaults.COMPRESSION_LEVEL : 1
19:03:42.999 INFO  ValidateSamFile - Defaults.CREATE_INDEX : false
19:03:42.999 INFO  ValidateSamFile - Defaults.CREATE_MD5 : false
19:03:42.999 INFO  ValidateSamFile - Defaults.CUSTOM_READER_FACTORY : 
19:03:42.999 INFO  ValidateSamFile - Defaults.EBI_REFERENCE_SERVICE_URL_MASK : http://www.ebi.ac.uk/ena/cram/md5/%s
19:03:42.999 INFO  ValidateSamFile - Defaults.NON_ZERO_BUFFER_SIZE : 131072
19:03:42.999 INFO  ValidateSamFile - Defaults.REFERENCE_FASTA : null
19:03:43.000 INFO  ValidateSamFile - Defaults.SAM_FLAG_FIELD_FORMAT : DECIMAL
19:03:43.000 INFO  ValidateSamFile - Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
19:03:43.000 INFO  ValidateSamFile - Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
19:03:43.000 INFO  ValidateSamFile - Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
19:03:43.000 INFO  ValidateSamFile - Defaults.USE_CRAM_REF_DOWNLOAD : false
19:03:43.000 INFO  ValidateSamFile - Deflater JdkDeflater
19:03:43.000 INFO  ValidateSamFile - Inflater JdkInflater
19:03:43.000 INFO  ValidateSamFile - Initializing engine
19:03:43.000 INFO  ValidateSamFile - Done initializing engine
ERROR: Record 9762, Read name 20GAVAAXX100126:7:2:8126:115177, bin field of BAM record does not equal value computed based on alignment start and end, and length of sequence to which read is aligned
ERROR: Record 24466, Read name 20FUKAAXX100202:7:46:13035:77621, bin field of BAM record does not equal value computed based on alignment start and end, and length of sequence to which read is aligned
ERROR: Record 97940, Read name 20FUKAAXX100202:5:7:21464:86224, bin field of BAM record does not equal value computed based on alignment start and end, and length of sequence to which read is aligned
ERROR: Record 97955, Read name 20GAVAAXX100126:5:7:1291:122571, bin field of BAM record does not equal value computed based on alignment start and end, and length of sequence to which read is aligned
ERROR: Record 112212, Read name 20GAVAAXX100126:8:1:1429:129840, bin field of BAM record does not equal value computed based on alignment start and end, and length of sequence to which read is aligned
ERROR: Record 126595, Read name 20FUKAAXX100202:6:46:9311:1219, bin field of BAM record does not equal value computed based on alignment start and end, and length of sequence to which read is aligned
...
ERROR: Read name 20FUKAAXX100202:3:66:14857:5877, Mate not found for paired read
ERROR: Read name 20GAVAAXX100126:3:42:3602:56427, Mate not found for paired read
ERROR: Read name 20FUKAAXX100202:3:43:9956:132423, Mate not found for paired read
ERROR: Read name 20GAVAAXX100126:5:28:11981:2516, Mate not found for paired read
ERROR: Read name 20FUKAAXX100202:6:24:17640:150063, Mate not found for paired read
ERROR: Read name 20GAVAAXX100126:7:6:14036:136793, Mate not found for paired read
ERROR: Read name 20FUKAAXX100202:8:4:15388:58751, Mate not found for paired read
ERROR: Read name 20GAVAAXX100126:7:46:18099:135243, Mate not found for paired read
ERROR: Read name 20FUKAAXX100202:1:61:19609:50570, Mate not found for paired read
ERROR: Read name 20FUKAAXX100202:7:24:21385:130712, Mate not found for paired read
ERROR: Read name 20GAVAAXX100126:6:65:2547:48691, Mate not found for paired read
ERROR: Read name 20FUKAAXX100202:7:44:20236:184614, Mate not found for paired read
Maximum output of [100] errors reached.
19:03:47.020 INFO  ValidateSamFile - Shutting down engine
[March 9, 2017 7:03:47 PM EST] org.broadinstitute.hellbender.tools.picard.sam.ValidateSamFile done. Elapsed time: 0.07 minutes.
Runtime.totalMemory()=1674051584
Tool returned:
false
```

gspowley commented:
Addressed comments and tests are passing, back to @droazen.

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterDeflaterIntegrationTest.java:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        Assert.assertTrue(new IntelDeflater().load(null), "IntelInflater/IntelDeflater shared library was not loaded. " +
+                "This could be due to a configuration error, or your system might not support it.");
```
I think you need two separate test cases here: one that tries to load the Intel deflater, and another that tries to load the Intel inflater.

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterDeflaterIntegrationTest.java:
```diff
@@ -0,0 +1,148 @@
+import java.util.zip.Inflater;
+
+/**
+ * Test that it's possible to load libIntelInflater
```
"Test that it's possible to load libIntelInflater" -> "Test GATK integration with the Intel inflater and deflater"

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterDeflaterIntegrationTest.java:
```diff
@@ -0,0 +1,148 @@
+    public Object[][] makeJdkFlags() {
+        final List<Object[]> tests = new ArrayList<>();
+
+        tests.add(new Object[]{false, true});
```
Add a one-line comment here documenting what the booleans in the arrays represent

gspowley replied:
done

droazen commented on src/test/java/org/broadinstitute/hellbender/IntelInflaterDeflaterIntegrationTest.java:
```diff
@@ -0,0 +1,148 @@
+            inflater.inflate(result);
+            inflater.end();
+
+            Assert.assertEquals(input, result);
```
Would it be a good idea to clear the `result` array after each iteration?

gspowley replied:
Yes, just to make sure the test doesn't pass somehow using previous results. Done.

droazen commented:
Final review complete, back to @gspowley. A few trivial remaining TODOs in the test code, then we can merge this.

gspowley commented:
Finished TODOs and tests are passing, back to @droazen.

droazen commented:
Looks good @gspowley -- merging!
